### ISSUE TITLE
Head segmentation in real time?

### ISSUE THREAD

**Issue by chris-hndz** [google-ai-edge/mediapipe#3503]
Hi, I am looking for a solution to achieve head segmentation (hair with face segmentation) in real time on a webcam video. What do you recommend to achieve this with mediapipe? What segmentation or mediapipe tracking methods would be useful to achieve this? 

![HeadSegmentation](https://user-images.githubusercontent.com/6933154/178130250-63e004cb-7746-4b7d-9441-e5460e82cccf.jpg)

Thanks in advance!

**Comment by brucechou1983** [comment]
I don't think you could achieve this by just using MediaPipe public models. We've done something similar in my team. We basically train our own model and serve it by MediaPipe.

**Comment by LGuilhermeRodrigues** [comment]
I would try Selfie Segmentation to remove the background and then Face Mesh to cut the neck

**Comment by sureshdagooglecom** [comment]
Hi @chris-hndz ,
These models are not part of MediaPipe solutions currently and this is a feature request.

**Comment by chris-hndz** [comment]
Thanks for the responses and ideas. I understand this is more of an improvement proposal for Mediapipe.

**Comment by kuaashish** [comment]
@chris-hndz,

Image Segmenter should be able to do this with the right model. You should try https://storage.googleapis.com/mediapipe-models/image_segmenter/deeplab_v3/float32/1/deeplab_v3.tflite with MediaPipe tasks. Thank you

**Comment by chris-hndz** [comment]
@kuaashish Thanks for your idea.

We did some tests with the Image Segmentation online demo with the _Multi-class selfie segmentation model_ and we obtained reasonably good results for the segmentation mask of the face and hair. Maybe if we put both masks together we can get the head segmentation mask. But it still has a notorious delay.

<img width="480" alt="headSegmMediaPipe" src="https://github.com/google/mediapipe/assets/6933154/bb3778ba-e1a2-4da3-bc3a-36dd48ed46d3">

And when we tested the Face Landmark Detection online demo we got this:

<img width="480" alt="GridRender" src="https://github.com/google/mediapipe/assets/6933154/2765a757-17ca-4f39-a999-c8c906475f4b">

Here we see that the face landmarks limit well the lower part of the head.

Is there any way to merge the face and hair segmentation masks, and considering also the cutting the lower part of the face mask according to the face landmarks of the lower part of the face?

Thank you for your answer!

**Comment by kuaashish** [comment]
@chris-hndz,

You can use Multi-class selfie segmentation model and combine mask 1 (hair) and 3 (face-skin) as guided below:

https://developers.google.com/mediapipe/solutions/vision/image_segmenter#multiclass-model 

However, If you will be able to achieve the full head model that can be instructions to other users. Thank you

**Comment by kuaashish** [comment]
@chris-hndz,

There is no off-the-shelf solution to this requirement currently. Thank you

**Comment by chris-hndz** [comment]
Thanks @kuaashish for your reply.

I understand that there is no off-the-shelf solution to this requirement.
Could you perhaps give some guidance or an idea of what would be the way to achieve this?

Thank you in advance!

**Comment by akashvverma1995** [comment]
Hi @chris-hndz 

Sorry for the delayed response! I tried this issue from my end. I attached the code for your reference.

`MultiClass selfie segmentation model` segments the body into  6 different categories (Background, Hair, Body-Skin, Face-Skin, Clothes, Accessories). By filtering the mask to only include the Hair and Face-Skin categories, we get a perfect head segmentation mask.

```
import cv2
import numpy as np
import mediapipe as mp
from mediapipe.tasks import python
from mediapipe.tasks.python import vision

# Initialize Multiclass Selfie Segmentation Model
base_options = python.BaseOptions(model_asset_path='selfie_multiclass_256x256.tflite')
options = vision.ImageSegmenterOptions(
    base_options=base_options,
    output_category_mask=True,
    running_mode=vision.RunningMode.VIDEO)

segmenter = vision.ImageSegmenter.create_from_options(options)
cap = cv2.VideoCapture(0)

# Target classes: 1 (Hair) and 3 (Face Skin)
TARGET_CLASSES = [1, 3]

while cap.isOpened():
    success, frame = cap.read()
    if not success: break
        
    rgb_frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
    mp_image = mp.Image(image_format=mp.ImageFormat.SRGB, data=rgb_frame)
    
    timestamp_ms = int(cv2.getTickCount() / cv2.getTickFrequency() * 1000)
    segmentation_result = segmenter.segment_for_video(mp_image, timestamp_ms)
    
    if segmentation_result.category_mask is not None:
        mask_data = np.squeeze(segmentation_result.category_mask.numpy_view())
        
        head_mask = np.isin(mask_data, TARGET_CLASSES).astype(np.uint8)
        
        # Color the head green for visualization
        colored_mask = np.zeros_like(frame)
        colored_mask[head_mask == 1] = (0, 255, 0)
        
        # Blend original frame with the mask
        frame = cv2.addWeighted(frame, 1, colored_mask, 0.5, 0)
        
    cv2.imshow('Head Segmentation', frame)
    if cv2.waitKey(1) & 0xFF == ord('q'): break

cap.release()
cv2.destroyAllWindows()
segmenter.close()

```
Hope this helps! Let me know if you have any other question.

**Comment by chris-hndz** [comment]
Hi @akashvverma1995 !

Thank you for your reply and for sharing the code.
I'll test it to see if it works.

Best regards!

**Comment by akashvverma1995** [comment]
Hi @chris-hndz 

Please let me know if your issue has been resolved, and kindly close this issue if it is.
